pkgs/node/src/forkchoice.zig

```diff
@@ -310,6 +310,27 @@ pub const ForkChoice = struct {
         return false;
     }
 
+    /// Get all ancestor block roots from the current finalized block,
+    /// traversing backwards, and collecting all blocks with slot > previousFinalizedSlot.
+    /// Stops traversal when previousFinalizedSlot is reached or at genesis.
+    pub fn getAncestorsofFinalized(self: *Self, allocator: Allocator, currentFinalized: types.Root, previousFinalizedSlot: ?types.Slot) ![]types.Root {
+        var ancestors = std.ArrayList(types.Root).init(allocator);
+
+        var current_idx_or_null = self.protoArray.indices.get(currentFinalized);
+
+        while (current_idx_or_null) |current_idx| {
+            const current_node = self.protoArray.nodes.items[current_idx];
+            if (previousFinalizedSlot) |previous_finalized_slot| {
+                if (current_node.slot == previous_finalized_slot) {
+                    break;
+                }
+            }
+            try ancestors.append(current_node.blockRoot);
+            current_idx_or_null = current_node.parent;
+        }
+        return ancestors.toOwnedSlice();
+    }
+
```

Comment:
also do a followup PR to add & call a function to prune forkchoice for all slots < new finalized, you will also need to drop other similar arrays we maintain one to one with the nodes (deltas or other pls check in forkchoice) and also update the parent index mappings in node